### DISCUSSION TITLE
exploring-the-dotnet-8-preview-form-binding-in-minimal-apis/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#396] [Announcements]
# exploring-the-dotnet-8-preview-form-binding-in-minimal-apis/

In this post I discuss the new form binding support for minimal APIs, the current limitations and rough edges, and how to work with antiforgery tokens

https://andrewlock.net/exploring-the-dotnet-8-preview-form-binding-in-minimal-apis/

**Comment by naasking** [comment]
> That's because forms are inherently susceptible to Cross Site Request Forgery (CSRF) attacks, so you need to use anti-forgery tokens to protect yourself.

This isn't correct as written. Form-encoded submissions aren't any more vulnerable to CSRF than requests using JSON. CSRF is a confused deputy attack so the problem has always been the *browser* getting confused, regardless of the format of the data being sent. You can have CSRF vulnerabilities with JS code that sends JSON to the server but without any forms on the page at all.

**Comment by IvanFeric** [comment]
Actually, the checkbox behavior is not that strange at all. I would assume you would get similar error with regular controllers as well. The reason why it's failing is in the way the form was declared, i.e. because the checkbox was defined as:

```html
<input type="checkbox" name="isCompleted" />
```

That's why Razor would convert `@Html.CheckBoxFor(m => m.isCompleted)` (or `<input asp-for="@Model.isCompleted" />`) to the folowing HTML:

```html
<input name="isCompleted" type="checkbox" value="true" />
<input name="isCompleted" type="hidden" value="false" />
```

This way, if checkbox wasn't checked, the browser would send `isCompleted=false`, and if it was checked, it would send `isCompleted=true,false` which .NET binds to `true`,

It has always been the .NET way to manually set `value` of the checkbox and add an additional hidden field with the same name and the value `false`..

**Reply by naasking** [reply]
Exactly. You could also get around this by declaring the bool in your view model as nullable, eg. Nullable<bool>. Then if it wasn't checked, the value is just null, so you just check for  `== true` or `!= true`.

**Comment by rosdi** [comment]
About the `multipart/form-data` binding in .NET 8, unfortunately it does not support binding to arrays. I tried to create an email api where applications can POST to the api (along with the attachments) like so:

```cs
//This won't bind
app.MapPost("/mail/send", ([FromForm] string[] to, [FromForm] string[] cc, [FromForm] string title, IFormFileCollection attachments) =>
    {
		//will throw System.InvalidOperationException: TryParse method found on string with incorrect format. Must be a static method with format...


//This work, since 'to' and 'cc' is not array...
app.MapPost("/mail/send", ([FromForm] string to, [FromForm] string cc, [FromForm] string title, IFormFileCollection attachments) =>
    {

//I end up with this, have to parse the parameters manually, this accepts all the parameters including the binary files
app.MapPost("/mail/send", (IFormCollection emailForm) =>
    {
```

**Reply by rosdi** [reply]
Ok apparently this works 😊, both `to` and `cc` will correctly receive all the values which then can be converted to array or List...

```cs
//This works!, 'to' and 'cc' can have many values...
app.MapPost("/mail/send", ([FromForm] StringValues to, [FromForm] StringValues cc, [FromForm] string title, IFormFileCollection attachments) =>
    {
```
